3.1.10 and 3.1.11 fails to catch no-unused-vars

- Minimal testcase: https://github.com/PiPeep/babel-eslint-testcase
- Impacts: facebook/react#4032

@zpao @spicyj

Having the same issue. 

`no-unused-vars` actually partially works for this sample:

```js
var x = 10;

var y = 10;
y = 5;

// By default, unused arguments cause warnings.
(function(foo) {
    return 5;
})();

// Unused recursive functions also cause warnings.
function fact(n) {
    if (n < 2) return 1;
    return n * fact(n - 1);
}
```

Output:

```
   7:10  error  foo is defined but never used    no-unused-vars
  12:9   error  fact is defined but never used   no-unused-vars
```

--- a/index.js
+++ b/index.js
@@ -232,7 +232,10 @@ function monkeypatch() {
   referencer.prototype.VariableDeclaration = function(node) {
     if (node.declarations) {
       for (var i = 0; i < node.declarations.length; i++) {
-        checkIdentifierOrVisit.call(this, node.declarations[i].id);
+        var type = node.declarations[i].id.typeAnnotation;
+        if (type) {
+          checkIdentifierOrVisit.call(this, type);
+        }
       }
     }
     variableDeclaration.call(this, node);

--- a/test/non-regression.js
+++ b/test/non-regression.js
@@ -950,7 +950,7 @@ describe("verify", function () {
     it("array #9", function () {
       verifyAndAssertMessages([
           "let arr = [1, 2, 3];",
-          "let b = [for (e of arr) String(e)];"
+          "let b = [for (e of arr) String(e)]; b;"
         ].join("\n"),
         { "no-unused-vars": 1, "no-undef": 1 },
         []
@@ -1081,5 +1081,13 @@ describe("verify", function () {
         []
       );
     });
+  });
+
+  it("detects minimal no-unused-vars case #120", function () {
+    verifyAndAssertMessages(
+      "var unused;",
+      { "no-unused-vars": 1 },
+      [ "1:4 unused is defined but never used no-unused-vars" ]
+    )
   });
 });

